netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,305 @@
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import static io.netty.handler.codec.http2.Http2Error.PROTOCOL_ERROR;
+import static io.netty.handler.codec.http2.Http2Exception.connectionError;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2CodecUtil;
+import io.netty.handler.codec.http2.Http2ConnectionAdapter;
+import io.netty.handler.codec.http2.Http2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.handler.codec.http2.Http2Stream;
+import io.netty.util.ReferenceCountUtil;
+
+import java.util.ArrayDeque;
+import java.util.Map;
+import java.util.Queue;
+import java.util.TreeMap;
+
+/**
+ * <p>
+ * Implementation of a {@link Http2ConnectionEncoder} that dispatches all method call to
+ * another {@link Http2ConnectionEncoder}, except for when the maximum number of
+ * concurrent streams limit is reached.
+ * </p>
+ * <p>
+ * When this limit is hit, instead of rejecting any new streams this implementation
+ * buffers newly created streams and their corresponding frames. Once an active stream
+ * gets closed or the maximum number of concurrent streams is increased, this encoder will
+ * automatically try to empty its buffer and create as many new streams as possible.
+ * </p>
+ * <p>
+ * This implementation makes the buffering mostly transparent and can be used as a drop
+ * in replacement for {@link DefaultHttp2ConnectionEncoder}.
+ * </p>
+ */
+class BufferingHttp2ConnectionEncoder extends DecoratingHttp2ConnectionEncoder {
+
+  // Buffer for any streams and corresponding frames that could not be created
```

Comment:
FYI: Feel free to use JavaDoc here. It _is_ actually possible to generate JavaDoc with private fields. That can be useful to get familiar with a new codebase.